test/functional/client_side_encryption_prose_tests.js

```diff
@@ -0,0 +1,974 @@
+'use strict';
+
+const mongodb = require('../../index');
+const chai = require('chai');
+const expect = chai.expect;
+chai.use(require('chai-subset'));
+
+// Tests for the ClientEncryption type are not included as part of the YAML tests.
+
+// In the prose tests LOCAL_MASTERKEY refers to the following base64:
+
+// .. code:: javascript
+
+//   Mng0NCt4ZHVUYUJCa1kxNkVyNUR1QURhZ2h2UzR2d2RrZzh0cFBwM3R6NmdWMDFBMUN3YkQ5aXRRMkhGRGdQV09wOGVNYUMxT2k3NjZKelhaQmRCZGJkTXVyZG9uSjFk
+
+describe(
+  'Client Side Encryption Prose Tests',
+  { requires: { clientSideEncryption: true } },
+  function() {
+    const dataDbName = 'db';
+    const dataCollName = 'coll';
+    const dataNamespace = `${dataDbName}.${dataCollName}`;
+    const keyVaultDbName = 'admin';
+    const keyVaultCollName = 'datakeys';
+    const keyVaultNamespace = `${keyVaultDbName}.${keyVaultCollName}`;
+    const kmsProviders = {
+      aws: {
+        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
+        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
+      },
+      local: {
+        key: Buffer.from(
+          'Mng0NCt4ZHVUYUJCa1kxNkVyNUR1QURhZ2h2UzR2d2RrZzh0cFBwM3R6NmdWMDFBMUN3YkQ5aXRRMkhGRGdQV09wOGVNYUMxT2k3NjZKelhaQmRCZGJkTXVyZG9uSjFk',
+          'base64'
+        )
+      }
+    };
+
+    const noop = () => {};
+
+    let mongodbClientEncryption;
+    function getMongodbClientEncryption() {
+      mongodbClientEncryption =
+        mongodbClientEncryption || require('mongodb-client-encryption')(mongodb);
+      return mongodbClientEncryption;
+    }
+
+    describe('Data key and double encryption', function() {
+      // Data key and double encryption
+      // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      // First, perform the setup.
+      beforeEach(function() {
+        const mongodbClientEncryption = getMongodbClientEncryption();
+
+        // #. Create a MongoClient without encryption enabled (referred to as ``client``). Enable command monitoring to listen for command_started events.
+        this.client = this.configuration.newClient(
+          {},
+          { useNewUrlParser: true, useUnifiedTopology: true, monitorCommands: true }
+        );
+
+        this.commandStartedEvents = [];
+        this.client.on('commandStarted', e => {
+          if (!e.ismaster) {
+            this.commandStartedEvents.push(e);
```

Comment:
[these utility methods](https://github.com/mongodb/node-mongodb-native/blob/master/test/functional/apm_tests.js#L11-L23) might be useful to reuse.

Reply:
Those are tricky in that
a) I'm using a set for one of these,
b) It binds specifically to the object, and I have some tests that explicitly rely on being able to easily clear the events, which may make it tricky.

I kind of want to make like, an `EventTracker` class:

```ts
class EventTracker {
  constructor(client: MongoClient, eventName: string, options: { includeList?: string[], excludeList?: string[] });
  get events(): APMEvent[];
  clear(): void;
  teardown(): void;
}

...

before(function() {
  this.client = ...;
  this.commandStartedEvents = new EventTracker(this.client, 'commandStarted', { includeList: ['insert']});
});
```

Reply:
Is using a `Set` a good idea here? You would miss cases where we emitted multiple events erroneously, for instance. I'm not sure what you mean in part `b` above, but I trust you know best.

Reply:
for the first point on the `Set`: sets are pointer based. So we would technically miss cases if we emitted the **exact same** event (memory reference). Still the point is valid.

The thing about part B is that if you do `filterForCommands`, you pass in a `bucket`. That means that if you want to clear the events, you need to clear them on that specific bucket. In contrast, using a closure that explicitly calls `this.events.push` allows you to do things like `this.events = []` and have it still work.